aten/src/ATen/native/BlasKernel.cpp

```diff
@@ -166,4 +188,82 @@ AT_FORALL_SCALAR_TYPES_AND(BFloat16, INSTANTIATE);
 AT_FORALL_COMPLEX_TYPES(INSTANTIATE);
 #undef INSTANTIATE
 
+namespace blas_impl {
+#if AT_BUILD_WITH_BLAS()
+float dot_fast_path(int n, float* x, int incx, float* y, int incy) {
+  return sdot_(&n, x, &incx, y, &incy);
+}
+
+double dot_fast_path(int n, double* x, int incx, double* y, int incy) {
+  return ddot_(&n, x, &incx, y, &incy);
+}
+#endif
+
+template <typename scalar_t>
+scalar_t dot_naive(
+    int64_t n,
+    scalar_t* x,
+    int64_t incx,
+    scalar_t* y,
+    int64_t incy) {
+  int64_t i;
+  scalar_t sum = 0;
+  for (i = 0; i < n; i++) {
+    sum += x[i * incx] * y[i * incy];
+  }
+  return sum;
+}
+
+} // namespace blas_impl
+
+template <typename scalar_t>
+scalar_t dot_impl_floating(int64_t n, scalar_t* x, int64_t incx, scalar_t* y, int64_t incy)
+{
+  if (n == 1) {
+    incx = 1;
+    incy = 1;
+  }
+#if AT_BUILD_WITH_BLAS()
+        if ((n <= INT_MAX) && (incx <= INT_MAX) && (incy <= INT_MAX)) {
+          return blas_impl::dot_fast_path(n, x, incx, y, incy);
+        } else {
+          return blas_impl::dot_naive(n, x, incx, y, incy);
+        }
+#else
+        { return blas_impl::dot_naive(n, x, incx, y, incy); }
+#endif
+}
+
+template <typename scalar_t>
+scalar_t dot_impl(int64_t n, scalar_t* x, int64_t incx, scalar_t* y, int64_t incy) {
+  if (n == 1) {
+    incx = 1;
+    incy = 1;
+  }
+  return blas_impl::dot_naive(n, x, incx, y, incy);
+}
+
+template <>
+float dot_impl(int64_t n, float* x, int64_t incx, float* y, int64_t incy) {
+  return dot_impl_floating(n, x, incx, y, incy);
+}
+
+template <>
+double dot_impl(int64_t n, double* x, int64_t incx, double* y, int64_t incy) {
+  return dot_impl_floating(n, x, incx, y, incy);
+}
+
+#define INSTANTIATE_DOT_IMPL(scalar_t)  \
+  template scalar_t dot_impl<scalar_t>( \
+      int64_t n, scalar_t * x, int64_t incx, scalar_t * y, int64_t incy);
+INSTANTIATE_DOT_IMPL(uint8_t);
```

Comment:
Why do you need all the instantiations here? Should not dispatch instantiate all the necessary templates?

Reply:
It is in the spirit of how it is already in
https://github.com/pytorch/pytorch/blob/8d49d9def84b1e1af743a0307b52bfb53fbee16d/aten/src/ATen/native/BlasKernel.cpp

I am not exactly sure of the reason why it is in `.cpp` file.

Maybe @zasdfgbnm can shed some light.

Reply:
I believe you can `#define INSTANTIATE_DOT_IMPL(scalar_t, _)`, and then

```C++
AT_FORALL_SCALAR_TYPES_AND(BFloat16, INSTANTIATE);
```

Reply:
Thanks.Will try that.

Also can you tell if there was a reason to put it in a cpp. Curious

Reply:
Because in a cpp file, the compiler will not generate code for a template function unless you used it somewhere in that file. But you are not using it in the same file, and you used it somewhere else, and for this file you just want the compiler to generate code and export the symbol. For this case, you need manually instantiate it to make the compiler generate the code.

Reply:
@zasdfgbnm Thanks for the reply. Agreed that compiler will stamp the template only when it is called with particular type.
However my question is why not have `BlasKernel.h` which has the template, and then in `Blas.cpp` instantiating the template in 
```C++
  return AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::Half, self.scalar_type(), "dot", [&] {
    Tensor result = at::empty({}, self.options());
    result.fill_(dot_impl<scalar_t>(self.numel(), self.data_ptr<scalar_t>(), self.stride(0), other.data_ptr<scalar_t>(), other.stride(0)));
    return result;
  }
```

Apologies if I am missing out anything obvious :). Thanks for your time.

Reply:
Haha, now all internal builds are failing with 
```
Summary: 
stderr: caffe2/aten/src/ATen/native/BlasKernel.cpp:259:45: error: explicit instantiation of 'dot_impl<float>' that occurs after an explicit specialization has no effect [-Werror,-Winstantiation-after-specialization]
AT_FORALL_SCALAR_TYPES_AND2(Half, BFloat16, INSTANTIATE_DOT_IMPL);
```
etc